### PR TITLE
Enable companion object generation in the avrohugger-based source generator

### PR DESCRIPTION
**Pull request by cb372**
Sorry, another giant diff... 🙏 

The main point of this PR is to switch from using the `@service` macro annotation to source-generating the companion object for Avro, like we did for Protobuf. But unfortunately that requires quite a few related changes. I can try to split this into a few smaller PRs for easier review, but I think it probably makes sense to group them all together in a single PR.

### Use implicit resolution to choose marshallers for request/response types

Initially I hardcoded the marshallers to ScalaPB's protobuf marshallers for simplicity, but now the time has come to do it properly.

I implemented the same mechanism that we use in the macro annotation:

1. generate an import to pull the appropriate marshaller instances into scope based on the serialization type (Avro/AvroWithSchema/Protobuf)
2. generate a method descriptor that summons implicit marshallers for its request and response types

Example:

```scala
case class MyRequest(a: Int)
case class MyResponse(b: String)

trait MyService[F[_]] {
  def myMethod(req: MyRequest): F[MyResponse]
}

object MyService {
  import higherkindness.mu.rpc.internal.encoders.{avro or avrowithschema or proto}._
  val myMethodDescriptor = MethodDescriptor.newBuilder(
    implicitly[Marshaller[MyRequest]],
    implicitly[Marshaller[MyResponse]]
  ).setType(...)
  ...
}
```

### Remove most of the custom codec imports

Until now we have provided the option to import custom Avro4s `SchemaFor`/`Encoder`/`Decoder` instances for:
* Plain (untagged) BigDecimal, encoded as a byte array using a custom format
* BigDecimal tagged with precision and scale, encoded as an Avro `decimal`
* `java.time.LocalDateTime`, encoded as a long containing the epoch millis

And similarly for Protobuf, custom pbdirect `PBScalarValueWriter`/`PBScalarValueReader` instances for:
* (untagged) BigDecimal
* `LocalDate`, `LocalDateTime` and `Instant`

However, most of these custom instances are no longer needed.

#### Protobuf

Since we have switched from pbdirect to ScalaPB for protobuf serialization, there's no point in importing instances of pbdirect type classes.

Also, ScalaPB will never generate code containing `BigDecimal`, `LocalDate`, `LocalDateTime` or `Instant`, so there's no need to define codecs for them.

It's still possible for users to configure ScalaPB to use types such as `BigDecimal` or `LocalDate` in its generated code, using ScalaPB's [TypeMapper feature](https://scalapb.github.io/docs/customizations#custom-types), but there's no need for sbt-mu-srcgen to get involved in that.

**Conclusion:** we can remove all the custom codec imports for protobuf.

#### Avro: untagged BigDecimal

Mu's custom codec for plain (untagged) BigDecimals has been deprecated for over 3 years, for good reason.

If Avrohugger generates a BigDecimal, it means the original Avro type was a decimal, with a maximum precision and scale (e.g. `decimal(6, 2)`). If we throw away that precision and scale information and use Mu's raw BigDecimal codec, we will produce wire values that don't match the Avro schema. This is pretty dangerous.

I considered 3 options to deal with this:

1. Keep the deprecated codec and port it to Scala 3. I don't want to do this, for the reason outlined above.
2. Remove the deprecated codec. Serialization of BigDecimals would fall back to Avro4s's built-in codec, which blindly assumes certain values for scale and precision. Of course, those values might not match the values provided for this field in the Avro schema, so this is just as dangerous as option 1.
3. Remove support for raw BigDecimals entirely. Configure Avrohugger to always tag the BigDecimal with its maximum precision and scale, allowing us to reconstruct the `decimal` logical type and serialize appropriately.

I went for option 3, as I don't see any downside to tagging BigDecimals with their precision and scale to prevent information loss. Especially in Scala 3, where the opaque type is a bit more ergonomic than the shapeless-tagged type.

**Conclusion:** we can remove the custom codec import for Avro + untagged BigDecimal.

#### Avro: tagged BigDecimal

In Scala 3, the Avro4s type class instances for `TaggedDecimal` are defined in the `TaggedDecimal` companion object, so we don't need to import anything.

In Scala 2, we still need an import to bring the instances for `BigDecimal @@ (precision, scale)` into scope.

**Conclusion:** we still need the Avro + tagged BigDecimal import, but only for Scala 2.

#### Avro: java.time.LocalDateTime

The only datetime-related Avro logical types that Avrohugger supports are:
* `date` --> `java.time.LocalDate`
* `timestamp-millis` --> `java.time.Instant`

Avrohugger will never generate code containing a `java.time.LocalDateTime`, so there is no need for a custom codec for that type.

**Conclusion:** we can remove the Avro + javatime import.

### Extend the TaggedDecimal Scalafix rule

In the previous PR I implemented a Scalafix rule to rewrite `BigDecimal @@ (Nat._6, Nat._2)` to `TaggedDecimal[6, 2]`. But I didn't realise that a couple of other patterns can be generated, when precision and scale are >= 10.

I've updated the rule to handle those cases.

### Update and extend the scripted tests

* Switch all the Avro tests to use the Avrohugger generator, in preparation for removing the skeuomorph implementation
* Add checks on the contents of the generated Scala 2/Scala 3 files

## Remaining work

I think the only thing left to do in sbt-mu-srcgen is removal of the skeuomorph Avro generator.

Most of the remaining work is in the mu-scala repo:
* update the tests to use source generation instead of the macro annotation, so we can run the tests for Scala 3
* tidy-up and deletion (e.g. removing the BigDecimal codec mentioned above)
* update the docs and examples

### CODE REVIEW

**Review by juanpedromoreno — APPROVED** [review]
> Avro: untagged BigDecimal

> I went for option 3, as I don't see any downside to tagging BigDecimals with their precision and scale to prevent information loss. Especially in Scala 3, where the opaque type is a bit more ergonomic than the shapeless-tagged type.

Sounds good to me 👍 

Thanks @cb372! Impressive work 🙌

File: core/src/main/scala/higherkindness/mu/rpc/srcgen/avro/rewrites/ReplaceShapelessTaggedDecimal.scala [hunk 1]
```diff
@@ -27,15 +31,57 @@ class ReplaceShapelessTaggedDecimal extends SyntacticRule("ReplaceShapelessTagge
...
+        Patch.replaceTree(t, taggedDecimalType(precision, scale))
+
+      case t @ Type.Apply(
+        t"@@",
+        List(
+          t"scala.math.BigDecimal",
+          Type.Tuple(
+            List(
+              Type.Tuple(Type.Select(_, precisionNat1) :: Type.Select(_, precisionNat2) :: Nil),
+              Type.Tuple(Type.Select(_, scaleNat1) :: Type.Select(_, scaleNat2) :: Nil)
```
Comment (fedefernandez) [inline core/src/main/scala/higherkindness/mu/rpc/srcgen/avro/rewrites/ReplaceShapelessTaggedDecimal.scala:71]: We could have a scale of two digits and a precision of one. Not sure if we need to consider this here in this pattern matching. I can't remember the place, but we were like making a validation around that in the refactor.

```bash
scala> val bd = BigDecimal("0.000000000012345")
val bd: scala.math.BigDecimal = 1.2345E-11

scala> bd.scale
val res4: Int = 15

scala> bd.precision
val res5: Int = 5
```

**Review by fedefernandez — APPROVED** [review]
Outstanding work @cb372. I have only one doubt, but the approach and the implementation make sense to me.

File: core/src/main/scala/higherkindness/mu/rpc/srcgen/avro/rewrites/ReplaceShapelessTaggedDecimal.scala (hunk 1 above)
Comment (fedefernandez) [reply core/src/main/scala/higherkindness/mu/rpc/srcgen/avro/rewrites/ReplaceShapelessTaggedDecimal.scala:71]: Awesome, thanks for the explanation

**Pull request merged**